// axios-better-stacktrace/lib/cjs/axiosBetterStacktrace.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var patchedSym = Symbol('axiosBetterStacktrace.patched');
// toString used instead of instanceOf for detecting error type to prevent problem mentioned in issue #5
var isError = function (error) {
    return error instanceof Error || Object.prototype.toString.call(error) === '[object Error]';
};
var isAxiosError = function (error) {
    return isError(error) && error.isAxiosError;
};
var axiosMethods = [
    'request',
    'get',
    'delete',
    'head',
    'options',
    'post',
    'put',
    'patch',
];
var axiosBetterStacktrace = function (axiosInstance, opts) {
    if (opts === void 0) { opts = {}; }
    var _a = opts.errorMsg, errorMsg = _a === void 0 ? 'Axios Better Stacktrace' : _a;
    // do nothing if input does not look like an axios instance
    if (!axiosInstance || !axiosMethods.some(function (method) { return axiosInstance.hasOwnProperty(method); })) {
        return;
    }
    // avoid potential memory leaks if axios instance already patched
    if (axiosInstance[patchedSym]) {
        return;
    }
    var originalHandlers = {
        request: axiosInstance['request'],
        get: axiosInstance['get'],
        delete: axiosInstance['delete'],
        head: axiosInstance['head'],
        options: axiosInstance['options'],
        post: axiosInstance['post'],
        put: axiosInstance['put'],
        patch: axiosInstance['patch'],
    };
    // enhance original response error with a topmostError stack trace
    var responseErrorInterceptorId = axiosInstance.interceptors.response.use(function (response) {
        if (response.config && isError(response.config.topmostError)) {
            // remove topmostError to not clutter config and expose it to other interceptors down the chain
            delete response.config.topmostError;
        }
        return response;
    }, function (error) {
        if (isAxiosError(error) && error.config && isError(error.config.topmostError)) {
            error.originalStack = error.stack;
            error.stack = "".concat(error.stack, "\n").concat(error.config.topmostError.stack);
            delete error.config.topmostError;
        }
        throw error;
    });
    axiosMethods.forEach(function (method) {
        if (method in axiosInstance) {
            switch (method) {
                case 'request': {
                    var originalHandler_1 = axiosInstance[method];
                    axiosInstance[method] = function axiosBetterStacktraceMethodProxy(config) {
                        return originalHandler_1(__assign(__assign({}, (config || {})), { topmostError: new Error(errorMsg) }));
                    };
                    break;
                }
                case 'get':
                case 'delete':
                case 'head':
                case 'options': {
                    var originalHandler_2 = axiosInstance[method];
                    axiosInstance[method] = function axiosBetterStacktraceMethodProxy(url, config) {
                        return originalHandler_2(url, __assign(__assign({}, (config || {})), { topmostError: new Error(errorMsg) }));
                    };
                    break;
                }
                case 'post':
                case 'put':
                case 'patch': {
                    var originalHandler_3 = axiosInstance[method];
                    axiosInstance[method] = function axiosBetterStacktraceMethodProxy(url, data, config) {
                        return originalHandler_3(url, data, __assign(__assign({}, (config || {})), { topmostError: new Error(errorMsg) }));
                    };
                    break;
                }
            }
            if (!axiosInstance[patchedSym]) {
                axiosInstance[patchedSym] = true;
            }
        }
    });
    // ensure consumer of the plugin can restore original handlers and remove custom interceptor
    return function () {
        axiosInstance.interceptors.response.eject(responseErrorInterceptorId);
        Object.assign(axiosInstance, originalHandlers);
    };
};
exports.default = axiosBetterStacktrace;
